### PR TITLE
Add the heap data-structure

### PR DESCRIPTION
**Pull request by QGarchery**
This implements a heap as the data structure for the matching engine. The implementation uses a "basic" heap in the sense that its elements are simply stored in an array with the following rules for the indexes:

- they start from index `1` and end at the length of the array (included)
- index `0` is used for checking if elements are not in the heap
- from index `i`, go to the left child with `2 * i`
- from index `i`, go to the right child with `2 * i + 1`
- from index `i`, go to the parent with `i / 2`

There is an address as satellite data `id`, that can be used, for example, to retrieve whose account has the biggest amount. There is also a mapping `indexes` from addresses to indexes in the array, allowing us to access the account associated to a given address.

The only non-view, non-private function is the function `update`. According to the parameters it receives, it calls one of the private functions `insert`, `decrease`, `increase` or `remove`.

### CODE REVIEW

File: contracts/Heap.sol [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        uint256 index = _heap.indexes[_id];
+        if (index < _heap.accounts.length) return _heap.accounts[index].id;
+        else return address(0);
+    }
+
+    /// PRIVATE ///
+
+    /// @notice Sets `_index` in the `_heap` to be `_account`.
+    /// @dev The heap may lose its invariant about the order of the values stored.
+    /// @dev Only call this function with an index in the bounds of the array.
```
Comment (MerlinEgalite) [inline contracts/Heap.sol:79]: still don't this is necessary as it would revert

File: contracts/Heap.sol [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    /// @notice Moves an account up the heap until its value is smaller than the one of its parent.
+    /// @dev This functions restores the invariant about the order of the values stored when the account at `_index` is the only one with value greater than what it should be.
+    /// @param _heap The heap to modify.
+    /// @param _index The index of the account to move.
+    function shiftUp(Heap storage _heap, uint256 _index) private {
+        Account memory initAccount = _heap.accounts[_index - 1];
+        uint256 initValue = initAccount.value;
+        while (_index != 1 && initValue > _heap.accounts[_index / 2 - 1].value) {
```
Comment (MerlinEgalite) [inline contracts/Heap.sol:116]: did you try the bit shift for all mul and iv by 2?

File: contracts/Heap.sol [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    /// @param _index The index of the account to move.
+    function shiftDown(Heap storage _heap, uint256 _index) private {
+        uint256 size = _heap.accounts.length;
+        Account memory initAccount = _heap.accounts[_index - 1];
+        uint256 childIndex = _index * 2;
+        Account memory childAccount;
+
+        while (childIndex <= size) {
+            if (
+                childIndex + 1 <= size &&
```
Comment (MerlinEgalite) [inline contracts/Heap.sol:135]: could still be otpimized

**Review by MerlinEgalite — APPROVED** [review]
Small comments bur overall LGTM

**Comment by QGarchery** [comment]
After thinking about it, it should be made more general. For example the `remove`, `insert`, `increase` and `decrease` should be made internal.
I'll leave this PR like this for the moment

**Review by pakim249CAL — APPROVED** [review]

**Comment by QGarchery** [comment]
We should also provide a way to abstract the satellite data in the structure

**Comment by QGarchery** [comment]
We can also implement:
- #38 
- #36

**Review by Rubilmax — APPROVED** [review]

**Pull request merged**